### PR TITLE
C#: Generate source files from cshtml files in standalone

### PR DESCRIPTION
**Pull request by tamasvajk**
Supersedes https://github.com/github/codeql/pull/13722.

This PR adds cshtml/razor file handling to the standalone extractor. We're generating C# files to a temporary location from the view files with `csc` by running the razor source generators on all the cshtml/razor files in the repo. The generated source files are then extracted as normal source files.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp.Standalone/Razor.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            {
+                sw.WriteLine($"\n[{f}]");
+                var base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(f)); // TODO: this should be the relative path of the file.
+                sw.WriteLine($"build_metadata.AdditionalFiles.TargetPath = {base64}");
+            }
+        }
+
+        public IEnumerable<string> GenerateFiles(IEnumerable<string> cshtmls, IEnumerable<string> references, string workingDirectory)
+        {
+            // TODO: the below command might be too long. It should be written to a temp file and passed to csc via @.
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp.Standalone/Razor.cs:54]: The integration test on Windows is probably failing due to this issue (even though the error message doesn't exactly say that). 

```
Failed to generate source files from cshtml files: An error occurred trying to start process 'dotnet' with working directory 'C:\a\semmle-code\semmle-code\target\codeql-csharp-integration-tests\ql\csharp\ql\integration-tests\all-platforms\cshtml_standalone'. The filename or extension is too long.
```

File: csharp/extractor/Semmle.Extraction.CSharp.Standalone/DotNet.cs [hunk 2]
```diff
@@ -76,23 +68,33 @@ public bool AddPackage(string folder, string package)
...
-            return runtimes;
+            progressMonitor.LogInfo($"Found {artifact}s: {string.Join("\n", artifacts)}");
+            return artifacts;
+        }
+
+        public bool Exec(string execArgs)
+        {
+            // TODO: we might need to swallow the stdout of the started process to not pollute the logs of the extraction.
+            var args = $"exec {execArgs}";
+            return RunCommand(args);
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.Standalone/DotNet.cs:97]: This is what might be causing the failing integration test on windows.
According to the (limited information here) we probably need to avoid invoking the command as a "shell execute": https://stackoverflow.com/questions/19306194/process-launch-exception-filename-or-extension-is-too-long-can-it-be-caused-b

It would be worth trying to start the process in the same way as in the `GetListed` method (`UseShellExecute = false`) and then also fix the TODO comment and swallow the standard out by re-directing it.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        private static void GenerateAnalyzerConfig(IEnumerable<string> cshtmls, string analyzerConfigPath)
+        {
+            using var sw = new StreamWriter(analyzerConfigPath);
+            sw.WriteLine("is_global = true");
+
+            foreach (var f in cshtmls.Select(f => f.Replace('\\', '/')))
+            {
+                sw.WriteLine($"\n[{f}]");
+                var base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(f)); // TODO: this should be the relative path of the file.
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs:46]: Why does it need to be the relative path?

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 4]
```diff
@@ -131,6 +132,40 @@ public DependencyManager(string srcDir, IDependencyOptions options, ILogger logg
...
+
+                    // TODO: use SDK specified in global.json
+                    var sdk = new Sdk(dotnet).GetNewestSdk();
+                    if (sdk != null)
+                    {
+                        try
+                        {
+                            var razor = new Razor(sdk, dotnet, progressMonitor);
+                            razorWorkingDirectory = new TemporaryDirectory(ComputeTempDirectory(sourceDir.FullName, "razor"));
+                            var generatedFiles = razor.GenerateFiles(views, usedReferences.Keys, razorWorkingDirectory.ToString());
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:157]: nit: consider to use `razoeWorkingDirectory.FullName` instead of calling `ToString()`.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                sw.WriteLine($"build_metadata.AdditionalFiles.TargetPath = {base64}");
+            }
+        }
+
+        public IEnumerable<string> GenerateFiles(IEnumerable<string> cshtmls, IEnumerable<string> references, string workingDirectory)
+        {
+            var name = Guid.NewGuid().ToString("N").ToUpper();
+            var tempPath = Path.GetTempPath();
+            var analyzerConfig = Path.Combine(tempPath, name + ".txt");
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs:55]: nit: Consider using string interpolation instead of "+".

File: csharp/ql/integration-tests/all-platforms/cshtml_standalone/Files.expected [hunk 6]
```diff
@@ -0,0 +1,3 @@
+| Program.cs |
+| Views/Home/Index.cshtml |
+| _semmle_code_target_codeql_csharp_integration_tests_ql_csharp_ql_integration_tests_all_platforms_cshtml_standalone_Views_Home_Index_cshtml.g.cs |
```
Comment (michaelnebel) [inline csharp/ql/integration-tests/all-platforms/cshtml_standalone/Files.expected:3]: Do we get this long name because this is the name of the integration test folder (+ Views/Home/Index.cshtml)?

File: csharp/tools/tracing-config.lua [hunk 7]
```diff
@@ -150,6 +150,8 @@ function RegisterExtractorPack(id)
     end
 
     local windowsMatchers = {
+        CreatePatternMatcher({ '^semmle%.extraction%.csharp%.standalone%.exe$' },
```
Comment (michaelnebel) [inline csharp/tools/tracing-config.lua:153]: Why is this needed?

**Review by michaelnebel — COMMENTED** [review]
This looks really really good!
It must have taken lots of reading and trial and error to get it to work!
Only some minor comments and questions.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 8]
```diff
@@ -131,6 +132,40 @@ public DependencyManager(string srcDir, IDependencyOptions options, ILogger logg
                 progressMonitor.UnresolvedReference(r.Key, r.Value);
             }
 
+            var webViewExtractionOption = Environment.GetEnvironmentVariable("CODEQL_EXTRACTOR_CSHARP_STANDALONE_EXTRACT_WEB_VIEWS");
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:135]: I think in the long run we'll have a couple of configuration options, so it definitely makes sense to group them. At the same time, we should try to minimize the number of options as the extractor should "just work". I don't have full confidence in this cshtml to C# generation, so I wanted to hide it behind an undocumented feature flag, but eventually I don't think there's any reason to not have this run all the time.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        private readonly string sourceGeneratorFolder;
+        private readonly string cscPath;
+
+        public Razor(DotnetVersion sdk, DotNet dotNet, ProgressMonitor progressMonitor)
+        {
+            this.sdk = sdk;
+            this.progressMonitor = progressMonitor;
+            this.dotNet = dotNet;
+
+            sourceGeneratorFolder = Path.Combine(this.sdk.FullPath, "Sdks", "Microsoft.NET.Sdk.Razor", "source-generators");
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs:23]: It's just a best guess. It's working now on all platforms, but I don't think there's any guarantee that it works on previous versions or upcoming ones.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs (hunk 3 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Razor.cs:46]: That's what the dotnet CLI is doing. I think this change would have some runtime consequences, but I'm unsure what. Also, it's not something that we currently rely on in any of the checks.

File: csharp/ql/integration-tests/all-platforms/cshtml_standalone/Files.expected (hunk 6 above)
Comment (tamasvajk) [reply csharp/ql/integration-tests/all-platforms/cshtml_standalone/Files.expected:3]: Yes.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs (hunk 4 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:157]: Just realized that `TemporaryDirectory` has no `FullName`. `ToString()` calls the underlying `DirectoryInfo.FullName`.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:135]: That makes sense!
In any case, we could consider to do the environment variable read in the options class instead (this might make it easier to test the entire DependencyManager class at some point in time), but we don't have to do it now.

File: csharp/tools/tracing-config.lua (hunk 7 above)
Comment (michaelnebel) [reply csharp/tools/tracing-config.lua:153]: Great - that makes sense!

**Review by michaelnebel — APPROVED** [review]
Really good work @tamasvajk !!
